src/main/java/com/github/alisonli/historyplugin/services/DiffAnalyzerService.java

```diff
@@ -98,6 +102,38 @@ private static RevisionDiffMetadata getRevisionMetadata(String beforeContent, St
         return revisionMetadata;
     }
 
+    public static Map<Integer, RevisionDiffMetadata> getRevisionMetadataMap(Project project, VirtualFile root, FileHistoryUi logUi)
+            throws VcsException {
+        VcsLogStorage logStorage = logUi.getLogData().getStorage();
+        GitLogDiffHandler diffHandler = new GitLogDiffHandler(project);
+        Set<Integer> commits =
+                FileHistoryPaths.INSTANCE.getFileHistory(logUi.getDataPack()).getCommitsToPathsMap().keySet();
+        List<VcsCommitMetadata> commitMetadataList = ContainerUtil.map(commits,
+                id -> logUi.getLogData().getMiniDetailsGetter().getCommitData(id, Collections.singleton(id)));
+        commitMetadataList.sort(Comparator.comparing(VcsShortCommitDetails::getCommitTime));
+
+        Map<Integer, RevisionDiffMetadata> metadataMap = new HashMap<>();
+        // The very first commit will end up with no comparison since the first commit is used as the
+        // "before" to compare with the second commit
+        Hash firstHash = commitMetadataList.get(0).getId();
+        ContentRevision firstContentRevision = diffHandler.createContentRevision(logUi.getPathInCommit(firstHash), firstHash);
+        String firstContent = firstContentRevision.getContent();
+        metadataMap.put(logStorage.getCommitIndex(firstHash, root), getRevisionMetadata("", firstContent));
+
+        for (int i = 0; i < commitMetadataList.size() - 1; i++) {
+            Hash hash1 = commitMetadataList.get(i).getId();
+            Hash hash2 = commitMetadataList.get(i + 1).getId();
+            ContentRevision contentRev1 = diffHandler.createContentRevision(logUi.getPathInCommit(hash1), hash1);
+            ContentRevision contentRev2 = diffHandler.createContentRevision(logUi.getPathInCommit(hash2), hash2);
+            String beforeContent = contentRev1.getContent();
+            String afterContent = contentRev2.getContent();
+            RevisionDiffMetadata metadata = getRevisionMetadata(beforeContent, afterContent);
+            metadataMap.put(logStorage.getCommitIndex(hash2, root), metadata);
+        }
+
+        return metadataMap;
+    }
```

Comment:
This entire method is my biggest concern - it seems horribly unoptimized but surprisingly it runs really fast when I build the plugin and use the action...

There's a bit of work that needs to be done because each commit has an integer represented (known as a commit index) for saving memory. When I retrieve the set of commit indices for a file, they may or may not be ordered so I have a comparator for getting a commit's details and using that to get the commit time to create a sorted commit list.

This method gets called by my `ShowDiffMetadataAction#performAction` to obtain a map of commits to revision diff metadata (contains number of documentation changes, newline changes, etc.). 
Actually, the whole map isn't needed if you look at what the result is used for in  `ShowDiffMetadataAction#performAction`. It takes the `RevisionMetadata` object for the one commit ID that it's interested in.

I'm wondering what you think the best way to refactor this is in your opinion? One way could be to store the result of this map somewhere so that it doesn't get computed every time the action is clicked. But the question is where to store it? And if this class should be the one storing it, then is it better for this class to not be static?